File: espresso-server/app/src/test/java/io/appium/espressoserver/test/helpers/w3c/adapter/AndroidKeyEventTest.java
package io.appium.espressoserver.test.helpers.w3c.adapter;

import android.os.SystemClock;
import android.view.KeyEvent;
import android.view.MotionEvent;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import java.util.LinkedList;
import java.util.List;

import androidx.test.espresso.UiController;
import io.appium.espressoserver.lib.handlers.exceptions.AppiumException;
import io.appium.espressoserver.lib.helpers.w3c.adapter.espresso.AndroidKeyEvent;
import io.appium.espressoserver.lib.helpers.w3c.adapter.espresso.EspressoW3CActionAdapter;
import io.appium.espressoserver.lib.helpers.w3c.dispatcher.KeyDispatch;
import io.appium.espressoserver.lib.helpers.w3c.dispatcher.W3CKeyEvent;

import static android.view.KeyEvent.ACTION_DOWN;
import static android.view.KeyEvent.ACTION_UP;
import static android.view.KeyEvent.KEYCODE_B;
import static android.view.KeyEvent.KEYCODE_C;
import static android.view.KeyEvent.KEYCODE_DEL;
import static android.view.KeyEvent.META_ALT_MASK;
import static android.view.KeyEvent.META_SHIFT_MASK;
import static io.appium.espressoserver.lib.helpers.w3c.dispatcher.constants.NormalizedKeys.BACKSPACE;
import static org.junit.Assert.assertEquals;

@RunWith(RobolectricTestRunner.class)
public class AndroidKeyEventTest {

    public static class MockUiController implements UiController {
        public List<KeyEvent> keyEvents = new LinkedList<>();

        @Override
        public boolean injectMotionEvent(MotionEvent event) {
            return false;
        }

        @Override
        public boolean injectMotionEventSequence(Iterable<MotionEvent> events) {
            return false;
        }

        @Override
        public boolean injectKeyEvent(KeyEvent event) {
            this.keyEvents.add(event);
            return true;
        }

        @Override
        public boolean injectString(String str) {
            return false;
        }

        @Override
        public void loopMainThreadUntilIdle() {
            SystemClock.sleep(100);
        }

        @Override
        public void loopMainThreadForAtLeast(long millisDelay) {
            SystemClock.sleep(millisDelay);
        }
    }

    @Test
    public void keyUpNoop() throws AppiumException {
        // Dispatch an up event that has no corresponding down event and check that the state was unchanged
        W3CKeyEvent w3CKeyEvent = new W3CKeyEvent();
        w3CKeyEvent.setKey("A");
        MockUiController mockUiController = new MockUiController();
        assertEquals(mockUiController.keyEvents.size(), 0);
        (new AndroidKeyEvent(mockUiController)).keyUp(w3CKeyEvent);
        assertEquals(mockUiController.keyEvents.size(), 0);
    }

    @Test
    public void keyDown() throws AppiumException {
        MockUiController mockUiController = new MockUiController();

        // Dispatch a key event
        W3CKeyEvent w3CKeyEvent = KeyDispatch.getKeyEvent(new EspressoW3CActionAdapter(mockUiController), "B");
        w3CKeyEvent.setAltKey(true);
        w3CKeyEvent.setShiftKey(true);
        w3CKeyEvent.setCtrlKey(false);
        (new AndroidKeyEvent(mockUiController)).keyDown(w3CKeyEvent);

        // Check that one was dispatched
        assertEquals(mockUiController.keyEvents.size(), 1);

        // Check the right key event was dispatched
        KeyEvent keyEvent = mockUiController.keyEvents.get(mockUiController.keyEvents.size() - 1);
        keyEvent.getKeyCode();
        assertEquals(keyEvent.getKeyCode(), KEYCODE_B);
        assertEquals(keyEvent.getAction(), ACTION_DOWN);
        assertEquals(keyEvent.getMetaState(), META_SHIFT_MASK | META_ALT_MASK);
    }

    @Test
    public void keyDownAndUp() throws AppiumException {
        MockUiController mockUiController = new MockUiController();
        AndroidKeyEvent androidKeyEvent = new AndroidKeyEvent(mockUiController);

        // Key Down C
        W3CKeyEvent w3CKeyEvent = KeyDispatch.getKeyEvent(new EspressoW3CActionAdapter(mockUiController), "C");
        androidKeyEvent.keyDown(w3CKeyEvent);

        // Check that KEYCODE_C went down
        KeyEvent keyEvent = mockUiController.keyEvents.get(mockUiController.keyEvents.size() - 1);
        keyEvent.getKeyCode();
        assertEquals(keyEvent.getKeyCode(), KEYCODE_C);
        assertEquals(keyEvent.getAction(), ACTION_DOWN);

        // Release Key Code C and check that there's no keys down
        androidKeyEvent.keyUp(w3CKeyEvent);
        keyEvent = mockUiController.keyEvents.get(mockUiController.keyEvents.size() - 1);
        keyEvent.getKeyCode();
        assertEquals(keyEvent.getKeyCode(), KEYCODE_C);
        assertEquals(keyEvent.getAction(), ACTION_UP);
    }

    @Test
    public void getKeyCode() {
        assertEquals(AndroidKeyEvent.getKeyCode(BACKSPACE, 0), KEYCODE_DEL);
        assertEquals(AndroidKeyEvent.getKeyCode("A", 0), -1);
    }
}

File: espresso-server/app/src/test/java/io/appium/espressoserver/test/helpers/w3c/adapter/AndroidKeyEventTest.kt
package io.appium.espressoserver.test.helpers.w3c.adapter

import android.os.SystemClock
import android.view.KeyEvent
import android.view.MotionEvent
import androidx.test.espresso.UiController
import io.appium.espressoserver.lib.handlers.exceptions.AppiumException
import io.appium.espressoserver.lib.helpers.w3c.adapter.espresso.AndroidKeyEvent
import io.appium.espressoserver.lib.helpers.w3c.adapter.espresso.EspressoW3CActionAdapter
import io.appium.espressoserver.lib.helpers.w3c.adapter.espresso.keyCodeToEvent
import io.appium.espressoserver.lib.helpers.w3c.dispatcher.W3CKeyEvent
import io.appium.espressoserver.lib.helpers.w3c.dispatcher.constants.NormalizedKeys.BACKSPACE
import io.appium.espressoserver.lib.helpers.w3c.dispatcher.getKeyEvent
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.util.*

@RunWith(RobolectricTestRunner::class)
class AndroidKeyEventTest {
    class MockUiController : UiController {
        var keyEvents: MutableList<KeyEvent> = LinkedList()
        override fun injectMotionEvent(event: MotionEvent): Boolean {
            return false
        }

        override fun injectMotionEventSequence(events: Iterable<MotionEvent>): Boolean {
            return false
        }

        override fun injectKeyEvent(event: KeyEvent): Boolean {
            keyEvents.add(event)
            return true
        }

        override fun injectString(str: String): Boolean {
            return false
        }

        override fun loopMainThreadUntilIdle() {
            SystemClock.sleep(100)
        }

        override fun loopMainThreadForAtLeast(millisDelay: Long) {
            SystemClock.sleep(millisDelay)
        }
    }

    @Test
    @Throws(AppiumException::class)
    fun keyUpNoop() {
        // Dispatch an up event that has no corresponding down event and check that the state was unchanged
        val w3CKeyEvent = W3CKeyEvent()
        w3CKeyEvent.key = "A"
        val mockUiController = MockUiController()
        Assert.assertEquals(mockUiController.keyEvents.size.toLong(), 0)
        AndroidKeyEvent(mockUiController).keyUp(w3CKeyEvent)
        Assert.assertEquals(mockUiController.keyEvents.size.toLong(), 0)
    }

    @Test
    @Throws(AppiumException::class)
    fun keyDown() {
        val mockUiController = MockUiController()

        // Dispatch a key event
        val w3CKeyEvent = getKeyEvent(EspressoW3CActionAdapter(mockUiController), "B")
        w3CKeyEvent.isAltKey = true
        w3CKeyEvent.isShiftKey = true
        w3CKeyEvent.isCtrlKey = false
        AndroidKeyEvent(mockUiController).keyDown(w3CKeyEvent)

        // Check that one was dispatched
        Assert.assertEquals(mockUiController.keyEvents.size.toLong(), 1)

        // Check the right key event was dispatched
        val keyEvent = mockUiController.keyEvents[mockUiController.keyEvents.size - 1]
        keyEvent.keyCode
        Assert.assertEquals(keyEvent.keyCode.toLong(), KeyEvent.KEYCODE_B.toLong())
        Assert.assertEquals(keyEvent.action.toLong(), KeyEvent.ACTION_DOWN.toLong())
        Assert.assertEquals(keyEvent.metaState, KeyEvent.META_SHIFT_MASK or KeyEvent.META_ALT_MASK)
    }

    @Test
    @Throws(AppiumException::class)
    fun keyDownAndUp() {
        val mockUiController = MockUiController()
        val androidKeyEvent = AndroidKeyEvent(mockUiController)

        // Key Down C
        val w3CKeyEvent = getKeyEvent(EspressoW3CActionAdapter(mockUiController), "C")
        androidKeyEvent.keyDown(w3CKeyEvent)

        // Check that KEYCODE_C went down
        var keyEvent = mockUiController.keyEvents[mockUiController.keyEvents.size - 1]
        keyEvent.keyCode
        Assert.assertEquals(keyEvent.keyCode.toLong(), KeyEvent.KEYCODE_C.toLong())
        Assert.assertEquals(keyEvent.action.toLong(), KeyEvent.ACTION_DOWN.toLong())

        // Release Key Code C and check that there's no keys down
        androidKeyEvent.keyUp(w3CKeyEvent)
        keyEvent = mockUiController.keyEvents[mockUiController.keyEvents.size - 1]
        keyEvent.keyCode
        Assert.assertEquals(keyEvent.keyCode.toLong(), KeyEvent.KEYCODE_C.toLong())
        Assert.assertEquals(keyEvent.action.toLong(), KeyEvent.ACTION_UP.toLong())
    }

    @Test
    fun keyCode() {
        Assert.assertEquals(keyCodeToEvent(BACKSPACE, 0).toLong(), KeyEvent.KEYCODE_DEL.toLong())
        Assert.assertEquals(keyCodeToEvent("A", 0).toLong(), -1)
    }
}